### ISSUE TITLE
Interop in native-image

### ISSUE THREAD

**Issue by mtneug** [oracle/truffleruby#1853]
The documentation states that 

> TruffleRuby’s implementation of Java interop does not work with the GraalVM Native Image Generator’s limited support for runtime reflection.

Is this also true for cases like these:

```java
import org.graalvm.polyglot.Context;
import java.util.concurrent.Callable;

public class Main {
    public static class MyClass {
        public int               id    = 42;
        public String            text  = "42";
        public int[]             arr   = new int[]{1, 42, 3};
        public Callable<Integer> ret42 = () -> 42;
    }

    public static void main(String[] args) {
        try (Context context = Context.newBuilder()
                                      .allowAllAccess(true)
                                      .build()) {
            context.getPolyglotBindings().putMember("javaObj", new MyClass());
            boolean valid = context.eval("ruby",
                    "javaObj = Polyglot.import('javaObj')\n" +
                            "    javaObj[:id]         == 42"         +
                            " && javaObj[:text]       == '42'"       +
                            " && javaObj[:arr][1]     == 42"         +
                            " && javaObj[:ret42].call == 42")
                                   .asBoolean();
            assert valid == true;
        }
    }
}
```

This example is taken from the [embedding documentation](https://www.graalvm.org/docs/reference-manual/embed/). When I create a native-image running on macOS I get the following:

```sh
$ javac Main.java 
$ native-image --language:ruby Main
Build on Server(pid: 91261, port: 61058)
[main:91261]    classlist:   2,359.16 ms
[main:91261]        (cap):   2,898.40 ms
[main:91261]        setup:   4,358.45 ms
[main:91261]   (typeflow):  91,474.68 ms
[main:91261]    (objects):  63,845.68 ms
[main:91261]   (features):  14,068.61 ms
[main:91261]     analysis: 175,780.60 ms
[main:91261]     (clinit):   3,419.48 ms
10492 method(s) included for runtime compilation
[main:91261]     universe:   9,040.64 ms
[main:91261]      (parse):   8,095.22 ms
[main:91261]     (inline):  13,722.54 ms
[main:91261]    (compile):  58,872.20 ms
[main:91261]      compile:  85,985.94 ms
[main:91261]        image:  17,028.59 ms
[main:91261]        write:   3,466.59 ms
[main:91261]      [total]: 299,157.81 ms
$ ./main
[ruby] WARNING: could not determine TruffleRuby's home - the standard library will not be available - set --home= or use --log.level=CONFIG to see details
[ruby] WARNING: could not determine TruffleRuby's home - the standard library will not be available - set --home= or use --log.level=CONFIG to see details
Exception in thread "main" Unknown identifier: id (NameError)
        at <ruby> <top (required)>(Unnamed:2:37-158)
        at <ruby> parsing-request(Unknown)
        at org.graalvm.polyglot.Context.eval(Context.java:370)
        at Main.main(Main.java:20)
```

I'm wondering why this example does not work since I'm not dynamically loading Java classes from Ruby. Or is this happening under the hood? Is it maybe possible to implement this behavior (accessing Java fields and methods from Ruby) in another way? Is it maybe possible to provide a reflection configuration?

**Comment by eregon** [comment]
That documentation is outdated, I'll update it.
Host (Java) interop works now with native images, but they need a reflection configuration file as documented.

To get that example working, you need to pass this reflection configuration file:
```
[
  { "name": "Main$MyClass", "allPublicFields": true },
  { "name": "java.util.concurrent.Callable", "allPublicMethods": true }
]
```
as mentioned in https://www.graalvm.org/docs/reference-manual/embed/#build-native-images-from-polyglot-applications under `Configuring Native Host Reflection`.

Using the [native image agent](https://github.com/oracle/graal/blob/master/substratevm/CONFIGURE.md) should also work to generate the file automatically after a run on JVM.

> I'm wondering why this example does not work since I'm not dynamically loading Java classes from Ruby. Or is this happening under the hood?

`Java.type` is using reflection under the hood to get the class and list fields and methods.

**Comment by mtneug** [comment]
@eregon thank you! This works perfectly.

Should I leave this issue open until the documentation is updated?

**Comment by eregon** [comment]
Yes, I'll update the documentation.

**Comment by eregon** [comment]
Fixed in 927e1dc88bc018afde97e5f153e0ee02c1bb0300.
Thank you for the issue, I hope it's clearer now.
